File: kotlin/200_Number_of_Islands.kt
class Solution {
    fun numIslands(grid: Array<CharArray>): Int {
        var count = 0;
        for (i in 0 until grid.size){
            for (j in 0 until grid[i].size){
                if (grid[i][j] == '1') {
                    count++;
                    BFS(grid, i, j);
                }
            }
        }
        
        return count;
    }

    fun BFS(grid: Array<CharArray>, i: Int, j: Int) {
        if (
            i < 0 
            || i >= grid.size
            || j < 0 
            || j >= grid[i].size
            || grid[i][j] == '0'
        ) {
            return;
        }
        
        grid[i][j] = '0';
        BFS(grid, i + 1, j);
        BFS(grid, i - 1, j);
        BFS(grid, i, j + 1);
        BFS(grid, i, j - 1);
    }
}

File: ts/200_Number_of_Islands.ts
function numIslands(grid: string[][]): number {
    let count = 0;
    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[i].length; j++) {
            if (grid[i][j] == '1') {
                count++;
                BFS(grid, i, j);
            }
        }
    }
    return count;
};

function BFS(grid: string[][], i: number, j: number) {
    if (
        i < 0 || i >= grid.length
        || j < 0 || j >= grid[i].length
        || grid[i][j] == "0"
    ) {
        return;
    }

    grid[i][j] = "0";
    BFS(grid, i + 1, j);
    BFS(grid, i - 1, j);
    BFS(grid, i, j + 1);
    BFS(grid, i, j - 1);
}

